drivers/sensor/icm20948/icm20948.c

```diff
@@ -0,0 +1,706 @@
+
+	int ret = i2c_reg_update_byte_dt(&cfg->i2c, REG_BANK_SEL, reg_bank_mask, bank << 4);
+
+	if (ret) {
+		LOG_ERR("Error switching banks");
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline void icm20948_temp_c(int32_t in, int32_t *out_c, uint32_t *out_uc)
+{
+	/* Offset by 21 degrees Celsius */
+	int64_t in100 = (in * 100) + (ROOM_TEMP_OFFSET_DEG * TEMP_SENSITIVITY_X100);
+
+	/* Whole celsius */
+	*out_c = in100 / TEMP_SENSITIVITY_X100;
+
+	/* Micro celsius */
+	*out_uc = ((in100 - (*out_c) * TEMP_SENSITIVITY_X100) * INT64_C(1000000)) /
+		  TEMP_SENSITIVITY_X100;
+}
+
+static void icm20948_convert_temp(struct sensor_value *val, int16_t raw_val)
+{
+	icm20948_temp_c((int32_t)raw_val, &val->val1, &val->val2);
+}
+
+static inline void icm20948_accel_mss(int64_t sensitivity, int32_t in, int32_t *out, int32_t *u_out)
+{
+	/* Convert to micrometers/s^2 */
+	int64_t in_ms = in * SENSOR_G;
+
+	/* meters/s^2 whole values */
+	*out = in_ms / (sensitivity * 1000000LL);
+
+	/* micrometers/s^2 */
+	*u_out = (in_ms - (*out * sensitivity * 1000000LL)) / sensitivity;
+};
+
+static void icm20948_convert_accel(const struct icm20948_config *cfg, int32_t raw_accel_value,
+				   struct sensor_value *output_value)
+{
+	int64_t sensitivity = 0; /* value equivalent for 1g */
+
+	switch (cfg->accel_fs) {
+	case ACCEL_FS_SEL_2G:
+		sensitivity = 16384;
+		break;
+	case ACCEL_FS_SEL_4G:
+		sensitivity = 8192;
+		break;
+	case ACCEL_FS_SEL_8G:
+		sensitivity = 4096;
+		break;
+	case ACCEL_FS_SEL_16G:
+		sensitivity = 2048;
+		break;
+	}
+
+	icm20948_accel_mss(sensitivity, raw_accel_value, &output_value->val1, &output_value->val2);
+}
+
+static int icm20948_convert_gyro(struct sensor_value *val, int16_t raw_val, uint8_t gyro_fs)
+{
+	int64_t gyro_sensitivity_10x = 0; /* value equivalent for 10x gyro reading deg/s */
+
+	switch (gyro_fs) {
+	case GYRO_FS_250:
+		gyro_sensitivity_10x = 1310;
+		break;
+	case GYRO_FS_500:
+		gyro_sensitivity_10x = 655;
+		break;
+	case GYRO_FS_1000:
+		gyro_sensitivity_10x = 328;
+		break;
+	case GYRO_FS_2000:
+		gyro_sensitivity_10x = 164;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	int64_t in10_rads = (int64_t)raw_val * SENSOR_PI * 10LL;
+
+	/* Whole rad/s */
+	val->val1 = (int32_t)(in10_rads / (gyro_sensitivity_10x * 180LL * 1000000LL));
+
+	/* microrad/s */
+	val->val2 = (int32_t)((in10_rads - (val->val1 * gyro_sensitivity_10x * 180LL * 1000000LL)) /
+			      (gyro_sensitivity_10x * 180LL));
+
+	return 0;
+}
+
+static inline void icm20948_mag_gauss(int32_t in, int32_t *out, int32_t *u_out)
+{
+
+	int32_t k_magnetic_flux_sensitivity_x100 = 15;
+
+	int64_t in100_gauss = in * 1000000LL; /* 1 GAUSS equals 100 uT */
+
+	*out = (int32_t)(in100_gauss / (k_magnetic_flux_sensitivity_x100 * 1000000LL));
+
+	*u_out = (int32_t)(in100_gauss - (*out * k_magnetic_flux_sensitivity_x100 * 1000000LL)) /
+		 k_magnetic_flux_sensitivity_x100;
+};
+
+static int icm20948_convert_mag(struct sensor_value *val, int16_t raw_val)
```

Comment:
probably this wrapper function can be avoided.  Also over here i see dissecting of sensor_value struct, this is not necessary in my opinion.

Reply:
done